### ISSUE TITLE
Forbid to use `&` and `|` with booleans

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#1657]
# Rule request

## Thesis

Yesterday I saw this code: 

```python
my_result = (first > 0 & second != 0) | other is None
```

And I was 😮 

So, we need to check all operations around `&` and `|` and detect any booleans there:
- Explicit `True` and `False`
- Compares with `is` or `==` or `!=`
- Compares with `>`, `>=`, etc
- Logical operations `not`, `or`, `and`

If we can detect at least one of these operations, then we should raise a violation. 
Let's call it `BinaryAndBooleanMixupViolation`.

Note: other binary operations are still fine!
So, these lines should not raise:
- `first & second`
- `first | 10`
- etc

## Reasoning

This case indicates that a person confused `&` with `and` and `|` with `or`.
This can be the case if a person is comming from another language.

**Comment by TohDeKai** [comment]
Please assign to me thank you

**Comment by sobolevn** [comment]
@TohDeKai I would love to! Can you please share your ideas about possible solution? This mees like a complicated task to me!

**Comment by thammaji** [comment]
Hi, I would like to try working on this issue. This would be my first issue, so I'm not sure of a solution yet as I'm still going through the contribution tutorials. I'll update once I have an idea for a solution, but any suggestion is welcome too!

**Comment by sobolevn** [comment]
LGTM, @thammaji! Thanks a lot!

**Comment by thammaji** [comment]
Hi, was a bit busy over the past few weeks, but I've now gone through all of the contribution tutorials and internal docs; am a little intimidated, but I think I mostly understand. 

I just wanted to make sure I'm going in the right direction here: 

For this rule, I would likely have to create a new AST-based visitor, probably inside compares.py or conditions.py, then I would add a violation BitwiseAndBooleanMixupViolation to best_practices.py where it describes the violation, and the visitor should have logic to raise that violation. Also I'm not sure if I missed this in the tutorial but how is the code for the violation determined?

**Comment by sobolevn** [comment]
Looks like it! 👍

**Comment by thammaji** [comment]
Hey, I'm making pretty good progress. I wrote my visitor in /visitors/ast/compares.py and I'm constantly failing 
0:1      WPS202 Found too many module members: 8 > 7. I'm not really sure what this means, it triggers even if my new visitor class is empty? Help with this would be much appreciated!

Additionally currently my approach to the rule is simply to disallow usage of bitwise compartors with anything other than a constant or a variable, but I think will create false positive, like this should be valid.
@pytest.fixture(params=IS_COMPARES | EQUAL_COMPARES | OTHER_COMPARES)

I think I will change the logic to be conservative in the other way around (that is actually check for presence of boolean operators nearby).

**Comment by sobolevn** [comment]
>  I'm not really sure what this means, it triggers even if my new visitor class is empty?

Yes, it means that you have too many items in a module.

**Comment by thammaji** [comment]
Oh I see. I will move my visitor to operators.py then (it fits there better anyway).
